Initial gRPC server/framework implementation

aseovic opened the pull request:
Here is the PR for the initial gRPC server implementation, as we discussed.

Docs are still work in progress, but we should be able to finish them over the next few days, while you are reviewing the rest of the PR.

romain-grecourt commented on examples/grpc/common/pom.xml:
```diff
@@ -0,0 +1,146 @@
+          <plugin>
+            <groupId>org.xolstice.maven.plugins</groupId>
+            <artifactId>protobuf-maven-plugin</artifactId>
+            <version>0.5.1</version>
```
The plugin version need to be managed in the top level parent pom.

romain-grecourt commented on examples/grpc/common/pom.xml:
```diff
@@ -0,0 +1,146 @@
+            <artifactId>protobuf-maven-plugin</artifactId>
+            <version>0.5.1</version>
+            <configuration>
+              <protocArtifact>com.google.protobuf:protoc:3.5.1-1:exe:${os.detected.classifier}</protocArtifact>
```
If this configuration is common to all grpc example, then it should go under pluginManagement in exmaples/grpc/pom.xml

thegridman replied:
It is also used by `config/etcd/pom.xml` so I'll move it to the top level pom where the plugin is declared.

romain-grecourt commented on examples/grpc/opentracing/pom.xml:
```diff
@@ -0,0 +1,118 @@
+                        <resource>
+                            <targetPath>/</targetPath>
+                            <directory>${project.build.directory}</directory>
+                            <include>${project.build.finalName}-fat.jar</include>
```
What's the "-fat" suffix ? We don't do fat jars in Helidon and I don't see anything that builds a jar ending with "-far", what am I missing ?

thegridman replied:
Apart from the fact that this example creates a gRPC server the majority of the way it works was copied almost verbatim from the `examples/webserver/opentracing/pom.xml` including the bit you are questioning. I can only assume that it builds some sort of fat jar to put into the Docker image.

romain-grecourt replied:
This example seems broken, there is nothing in the pom.xml that builds a `-fat.jar`.

romain-grecourt commented on examples/grpc/pom.xml:
```diff
@@ -0,0 +1,42 @@
+
+    <groupId>io.helidon.examples.grpc</groupId>
+    <artifactId>helidon-examples-grpc-project</artifactId>
+    <name>Helidon gRPC Examples Parent</name>
```
Parent poms don't have a "Parent" name.

romain-grecourt commented on examples/grpc/security-outbound/pom.xml:
```diff
@@ -0,0 +1,101 @@
+           <version>${project.version}</version>
+           <exclusions>
+               <exclusion>
+                   <groupId>com.google.code.findbugs</groupId>
```
Why the exclusion ? This should rather be excluded where it's coming from, or marked optional if this is a helidon module.

thegridman replied:
This came from copying things from `examples/security/attribute-based-access-control/pom.xml` which has the same exclusion. I'm happy to remove it though.

romain-grecourt replied:
@tomas-langer Can you comment on that ?

tomas-langer replied:
This is a left-over from the time Google login was part of the security bundle (if I remember correctly).
They had a runtime dependency on this library, which I do not like, so I excluded it.

romain-grecourt commented on grpc/client/pom.xml:
```diff
@@ -0,0 +1,141 @@
+            <extension>
+                <groupId>kr.motd.maven</groupId>
+                <artifactId>os-maven-plugin</artifactId>
+                <version>1.5.0.Final</version>
```
This version needs to be managed in the top level parent pom.

romain-grecourt commented on grpc/client/pom.xml:
```diff
@@ -0,0 +1,141 @@
+            <plugin>
+                <groupId>org.xolstice.maven.plugins</groupId>
+                <artifactId>protobuf-maven-plugin</artifactId>
+                <version>0.5.1</version>
```
The version needs to be managed in the Helidon parent pom.

romain-grecourt commented on grpc/core/pom.xml:
```diff
@@ -0,0 +1,85 @@
+    </dependency>
+    <dependency>
+        <groupId>javax.json</groupId>
+        <artifactId>javax.json-api</artifactId>
```
Needs to be changed to "jakarta.json:jakarta.json-api"

thegridman replied:
The Helidon webserver still uses this same `javax.json` dependency and `jakarta.json:jakarta.json-api` is not yet in the Helidon top level `pom.xml`. I'm happy to switch but I'd rather wait until the webserver has changed and the dependency has been added rather than me do it and get merge conflicts. Or I can do the change and we'll sort out conflicts when we need to.

romain-grecourt replied:
We merged a PR a few days that changed, see the top level pom in the master branch.

romain-grecourt commented on grpc/server/pom.xml:
```diff
@@ -0,0 +1,213 @@
+            <extension>
+                <groupId>kr.motd.maven</groupId>
+                <artifactId>os-maven-plugin</artifactId>
+                <version>1.5.0.Final</version>
```
This version needs to be managed in the top level parent pom.

romain-grecourt commented on grpc/server/pom.xml:
```diff
@@ -0,0 +1,213 @@
+            <plugin>
+                <groupId>org.xolstice.maven.plugins</groupId>
+                <artifactId>protobuf-maven-plugin</artifactId>
+                <version>0.5.1</version>
```
This version needs to be managed in the top level parent pom.

romain-grecourt commented on grpc/server/pom.xml:
```diff
@@ -0,0 +1,213 @@
+                <artifactId>protobuf-maven-plugin</artifactId>
+                <version>0.5.1</version>
+                <configuration>
+                    <protocArtifact>com.google.protobuf:protoc:3.5.1-1:exe:${os.detected.classifier}</protocArtifact>
```
This plugin configuration should go under pluginManagement under grpc/pom.xml

romain-grecourt commented on security/integration/grpc/pom.xml:
```diff
@@ -0,0 +1,177 @@
+          <extension>
+            <groupId>kr.motd.maven</groupId>
+            <artifactId>os-maven-plugin</artifactId>
+            <version>1.5.0.Final</version>
```
This version needs to be managed in the top level parent pom.

romain-grecourt commented on security/integration/grpc/pom.xml:
```diff
@@ -0,0 +1,177 @@
+          <plugin>
+              <groupId>org.apache.maven.plugins</groupId>
+              <artifactId>maven-compiler-plugin</artifactId>
+              <version>3.7.0</version>
```
The version is redundant here, it's already managed in the top level pom.

romain-grecourt commented on security/integration/grpc/pom.xml:
```diff
@@ -0,0 +1,177 @@
+              <artifactId>maven-compiler-plugin</artifactId>
+              <version>3.7.0</version>
+              <configuration>
+                  <source>1.8</source>
```
Unless I'm mistaking source, target are set to 1.8 by default, is that actually needed ?

romain-grecourt commented on security/integration/grpc/pom.xml:
```diff
@@ -0,0 +1,177 @@
+          <plugin>
+            <groupId>org.xolstice.maven.plugins</groupId>
+            <artifactId>protobuf-maven-plugin</artifactId>
+            <version>0.5.1</version>
```
This plugin version should be managed in the top level pom.

thegridman commented:
Regarding the comment about `@author` tags, I'm happy to go through and remove them. There are Helidon classes with them in with various formats of author name so we had assumed it was ok to leave ours in.

Regarding the Spotify plugin - I think that this is just the Open Tracing example which builds a Docker image and as I said in another comment (and you also alluded to) was copied from the webserver example. Happy to change this to something else when we know what that is going to be.

romain-grecourt commented:
I'm aware of the few `@author` usages in the workspace (I did grep for it before I made this comment),  let's wait for some input from the team before acting on that.
No need to do anything with the Spotify plugin for now.

aseovic commented:
FWIW, my take on `@author` tags is that I'd prefer to leave them in -- it makes it easier to find out who created and worked on a file than going through Git history (especially with multiple forks, PRs, etc.)

That said, if we decide we shouldn't use them, I'm fine with removing them as well.

tomas-langer commented:
`@author` tag will be removed from the project (except for classes copied from other sources, where we copy them verbatim). 
We can use `git` to find authors if really needed.

tomas-langer commented on examples/grpc/security-abac/src/main/resources/application.conf:
```diff
@@ -0,0 +1,91 @@
+# limitations under the License.
+#
+
+grpc {
```
Please use yaml

aseovic commented:
@tomas-langer We've made all the changes discussed last Thursday, apart from using `ServiceLoader` to add vendor metrics dynamically if `grpc-metrics` is on the class path (we are still registering them manually in `MetricsSupport`).

I'll try to make that change tomorrow or Friday, but we should probably schedule another review meeting some time next week.

tomas-langer commented:
Hi,
 I have gone through the latest versions:

- Outbound security should not require additional classes to be defined in the application using it - there is a `BasicAuthCallCredentials` defined in both examples, even though the metadata uses same name and value as HTTP  basic authentication. This type of security should work without additional classes (it may require configuring the `SecurityContext` on the client, until we have the `Context` pull request merged
- the class `InterceptorPriority` defines an enum, which is quite restrictive. We usually use `int` for priority definition. Please have a  look at `Prioritized` in the latest master, as that describes how we want to handle priorities. If you feel that an enum is fully sufficient (e.g. there is no need to order interceptors within each enum value) - then you are probably fine. Nevertheless if you use the priority (either `Prioritized` or `@Priority), you can also use the `HelidonServiceLoader` that loads all the services and orders them by priority (with customization support)
- in `Server.java` in grpc/basics example, line 82 - remove `metrics and`
- in Zipkin example, I would use configuration: ```
tracing:
  service: "grpc-server"
```, and then use `Tracer tracer = TracerBuilder.create(config.get("tracing")).build();` in `ZipkinExampleMain` - the `localhost:9411` is the documented default for zipkin. This also gives you a way to easily change the endpoint using configuration, rather than changing the soruces.

tomas-langer commented:
Some of the changes we discussed (such as the usage of service loader) can be left after merge.
We should create a follow up task with any changes that are OK to be done later (e.g. a functionality is missing, yet will not break backward compatibility once added)

thegridman commented:
I have removed the `BasicAuthCallCredentials` class from the gRPC security examples. We actually had proper support for security integrated with the Helidon providers but we had not changed the examples to use it. 

Regarding the interceptor priority - I agree that an enum is too restrictive and I had in fact already changed that but it was on the branch that I have been doing the gRPC MP work in. I have now applied those changes to this PR too.

tomas-langer commented:
Please update the documentation to clearly mark this as experimental feature.
Please resolve the issue with pom.xml and merge.
Once we are confident this is finalized, we can remove that "experimental" flag.

aseovic commented:
@tomas-langer Looks like JK addressed the issues we discussed. 

If you can get it marged per our discussion today, I may be able to submit metrics PR over the weekend or early next week (some of which would be very useful for Cloudlogic demo in Rome, which is why I'm kinda eager to get it in).
